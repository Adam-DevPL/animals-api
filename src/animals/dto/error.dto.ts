import { ApiProperty } from '@nestjs/swagger';

export class ErrorDto {
  @ApiProperty({
    description: 'status of the error response',
    type: 'Number',
    example: '500',
  })
  status: number;

  @ApiProperty({
    description: 'error message',
    type: 'String',
    example: 'Internal server error',
  })
  message: string;
}
