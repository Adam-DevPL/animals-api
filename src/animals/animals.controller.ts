import { Body, Controller, Get, Param, Post, Put } from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiInternalServerErrorResponse,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiOperation,
  ApiParam,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { ParamsWithId } from 'src/validations/id.validator';
import { AnimalsService } from './animals.service';
import { AnimalDto } from './dto/animal.dto';

@ApiTags('animals')
@Controller('animals')
export class AnimalsController {
  constructor(private readonly animalsService: AnimalsService) {}

  @Get('/all')
  @ApiOperation({ summary: 'Get all animals from the api' })
  @ApiResponse({ status: 200, description: 'Get all data' })
  @ApiResponse({ status: 500, description: 'Internal server error' })
  async getAllAnimals() {
    return this.animalsService.findAll();
  }

  @Get('animal/:id')
  @ApiParam({
    name: 'id',
    description: 'Gets the Animal id',
  })
  @ApiOperation({ summary: 'Get an animal with a given id' })
  @ApiOkResponse({ description: 'The resources were returned successfully' })
  @ApiBadRequestResponse({ description: 'Incorrect id number' })
  @ApiNotFoundResponse({ description: 'Animal not found' })
  @ApiInternalServerErrorResponse({ description: 'Internal server error' })
  async getAnimal(@Param() { id }: ParamsWithId) {
    return this.animalsService.findOne(id);
  }

  // @Post('/add')
  // async createAnimal(@Body() animal: AnimalDto) {
  //   return this.animalsService.create(animal);
  // }

  // @Put('animal/:id')
  // async updateAnimal(@Param() { id }: ParamsWithId, @Body() animal: AnimalDto) {
  //   return this.animalsService.update(id, animal);
  // }
}
