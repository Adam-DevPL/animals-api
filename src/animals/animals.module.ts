import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { RedisCacheModule } from 'src/redis-cache/redis-cache.module';
import { AnimalsController } from './animals.controller';
import { AnimalsService } from './animals.service';
import { Animal, AnimalSchema } from './schemas/animal.schema';

@Module({
  imports: [
    MongooseModule.forFeature([{ name: Animal.name, schema: AnimalSchema }]),
    RedisCacheModule,
  ],
  controllers: [AnimalsController],
  providers: [AnimalsService],
})
export class AnimalsModule {}
