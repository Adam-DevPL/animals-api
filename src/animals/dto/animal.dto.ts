import { Optional } from '@nestjs/common';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
  IsString,
  IsNotEmpty,
  IsEnum,
  IsOptional,
  IsArray,
  ValidateNested,
} from 'class-validator';
import { AnimalType } from 'src/types/animals.type';

export class AnimalDto {
  @ApiProperty({
    description: 'Animal name, can not be empty',
    type: 'string',
    example: 'Cat',
  })
  @IsString()
  @IsNotEmpty()
  animalName: string;

  @ApiProperty({
    description: `One of Animal types - see Enum`,
    enum: AnimalType,
    type: 'AnimalType',
    example: AnimalType.MAMMALS,
  })
  @IsNotEmpty()
  @IsEnum(AnimalType)
  type: AnimalType;

  @ApiPropertyOptional({
    description: 'Optional description',
    type: 'string',
    example: 'test description',
  })
  @IsOptional()
  @IsString()
  description?: string;
}

export class UpdateAnimalDto {
  @ApiPropertyOptional({
    description: 'Animal name',
    type: 'string',
    example: 'Cat',
  })
  @IsOptional()
  @IsString()
  animalName?: string;

  @ApiPropertyOptional({
    description: `One of Animal types - see Enum`,
    enum: AnimalType,
    type: 'AnimalType',
    example: AnimalType.MAMMALS,
  })
  @Optional()
  @IsEnum(AnimalType)
  type?: AnimalType;

  @ApiPropertyOptional({
    description: 'Optional description',
    type: 'string',
    example: 'test description',
  })
  @IsOptional()
  @IsString()
  description?: string;
}

export class AnimalDtoArray {
  @ApiProperty({ description: 'Array with animals', type: [AnimalDto] })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => AnimalDto)
  animals: AnimalDto[];
}
