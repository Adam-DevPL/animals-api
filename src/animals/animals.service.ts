import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
  NotFoundException,
} from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import { AnimalDto } from './dto/animal.dto';
import { Animal, AnimalDocument, AnimalWithId } from './schemas/animal.schema';

@Injectable()
export class AnimalsService {
  constructor(
    @InjectModel(Animal.name) private animalModel: Model<AnimalDocument>,
  ) {}

  async findAll() {
    try {
      return this.animalModel.find();
    } catch (err) {
      console.error(err);
      throw new InternalServerErrorException();
    }
  }

  async findOne(id: string) {
    try {
      const animal: AnimalDocument = await this.animalModel.findById({
        _id: new Types.ObjectId(id),
      });
      if (!animal) {
        throw new NotFoundException();
      }
      return animal as AnimalWithId;
    } catch (err) {
      console.error(err);
      if (err instanceof NotFoundException) {
        throw new NotFoundException();
      }
      throw new InternalServerErrorException();
    }
  }

  async addAnimals(animalsList: AnimalDto[]) {
    try {
      const allAnimals: AnimalWithId[] = await this.findAll();
      const alreadyExistAnimals = animalsList.filter((animal) =>
        allAnimals.some(
          ({ animalName, type }) =>
            animal.animalName === animalName && animal.type === type,
        ),
      );
      if (alreadyExistAnimals.length !== 0) {
        throw new BadRequestException('Animals already exist in database');
      }
      return await this.animalModel.insertMany(
        animalsList.map((animal) => ({ createdAt: new Date(), ...animal })),
      );
    } catch (err) {
      console.error(err);
      if (err instanceof BadRequestException) {
        throw new BadRequestException(err.message);
      }
      throw new InternalServerErrorException();
    }
  }
}
