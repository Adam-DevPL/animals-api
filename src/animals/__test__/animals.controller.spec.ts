import { BadRequestException, NotFoundException } from '@nestjs/common';
import { InternalServerErrorException } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { plainToInstance } from 'class-transformer';
import { validate } from 'class-validator';
import { AnimalType } from 'src/types/animals.type';
import { ParamsWithId } from 'src/validations/id.validator';
import { AnimalsController } from '../animals.controller';
import { AnimalsService } from '../animals.service';
import { AnimalDto, AnimalDtoArray } from '../dto/animal.dto';
import { Animal } from '../schemas/animal.schema';

const animalStub: Animal = {
  animalName: 'Human',
  type: AnimalType.MAMMALS,
  createdAt: new Date('2023-03-19T18:27:12.933Z'),
  description: 'test description',
};

const animalsList: AnimalDto[] = [
  {
    animalName: 'Elephant',
    type: AnimalType.MAMMALS,
  },
  {
    animalName: 'Lion',
    type: AnimalType.MAMMALS,
  },
];

describe('AnimalsController', () => {
  let animalsController: AnimalsController;
  let animalsService: AnimalsService;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      controllers: [AnimalsController],
      providers: [
        AnimalsService,
        {
          provide: AnimalsService,
          useValue: {
            findAll: jest.fn().mockResolvedValue([animalStub]),
            findOne: jest.fn().mockResolvedValue(animalStub),
            addAnimals: jest.fn().mockImplementation((animals) =>
              Promise.resolve([
                {
                  _id: 'id1',
                  createdAt: '2022-02-02',
                  ...animals[0],
                },
                {
                  _id: 'id2',
                  createdAt: '2022-02-02',
                  ...animals[1],
                },
              ]),
            ),
          },
        },
      ],
    }).compile();

    animalsController = moduleRef.get<AnimalsController>(AnimalsController);
    animalsService = moduleRef.get<AnimalsService>(AnimalsService);
  });

  describe('getAllAnimals', () => {
    it('should return an array of animals', async () => {
      //whem
      const result = await animalsController.getAllAnimals();

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw an error when there is something wrong with mongodb connection', async () => {
      jest.spyOn(animalsService, 'findAll').mockImplementation(() => {
        throw new InternalServerErrorException();
      });

      expect(animalsController.getAllAnimals()).rejects.toMatchSnapshot();
    });
  });

  describe('getAnimal', () => {
    it('should return an animal', async () => {
      // given
      const id: ParamsWithId = { id: '507f1f77bcf86cd799439011' }; // example of id, correct with mongodb policy

      //when
      const result = await animalsController.getAnimal(id);

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error when animal not found', async () => {
      //given
      jest.spyOn(animalsService, 'findOne').mockImplementation(() => {
        throw new NotFoundException();
      });

      //then
      await expect(
        animalsController.getAnimal({
          id: '507f1f77bcf86cd799439011',
        }),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error when id is not mongodb id', async () => {
      //given
      const incorrectId = { id: 123 };
      const myObject = plainToInstance(ParamsWithId, incorrectId);

      //when
      const errors = await validate(myObject);

      //then
      expect(errors.length).not.toBe(0);
      expect(JSON.stringify(errors)).toContain(`id must be a mongodb id`);
    });
  });

  describe('addAnimalsLsit', () => {
    it('should add list of animals', async () => {
      //given
      const animalDtoArray = {
        animals: animalsList,
      };
      //when
      const result = await animalsController.addAnimalsList(animalDtoArray);

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error - BadRequestException - when one of animals already exist', async () => {
      //given
      const animalDtoArray = {
        animals: animalsList,
      };
      jest.spyOn(animalsService, 'addAnimals').mockImplementation(() => {
        throw new BadRequestException('Animals already exist in database');
      });

      //then
      await expect(
        animalsController.addAnimalsList(animalDtoArray),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error when no array as an entry DTO- BadRequestException', async () => {
      //given
      const incorrectType = { type: 123 };
      const myObject = plainToInstance(AnimalDtoArray, incorrectType);

      //when
      const errors = await validate(myObject);

      //then
      expect(errors.length).not.toBe(0);
      expect(JSON.stringify(errors)).toContain(`animals must be an array`);
    });

    it('should throw Interal Servel Error Exception when something goes worng with mongodb connection', async () => {
      //given
      const animalDtoArray = {
        animals: animalsList,
      };
      jest.spyOn(animalsService, 'addAnimals').mockImplementation(() => {
        throw new InternalServerErrorException();
      });

      //then
      expect(
        animalsController.addAnimalsList(animalDtoArray),
      ).rejects.toMatchSnapshot();
    });
  });
});
