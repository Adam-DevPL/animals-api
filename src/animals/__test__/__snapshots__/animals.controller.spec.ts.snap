// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalsController addAnimalsLsit should add list of animals 1`] = `
[
  {
    "_id": "id1",
    "animalName": "Elephant",
    "createdAt": "2022-02-02",
    "type": "MAMMALS",
  },
  {
    "_id": "id2",
    "animalName": "Lion",
    "createdAt": "2022-02-02",
    "type": "MAMMALS",
  },
]
`;

exports[`AnimalsController addAnimalsLsit should throw Interal Servel Error Exception when something goes worng with mongodb connection 1`] = `[InternalServerErrorException: Internal Server Error]`;

exports[`AnimalsController addAnimalsLsit should throw error - BadRequestException - when one of animals already exist 1`] = `[BadRequestException: Animals already exist in database]`;

exports[`AnimalsController getAllAnimals should return an array of animals 1`] = `
[
  {
    "animalName": "Human",
    "createdAt": 2023-03-19T18:27:12.933Z,
    "description": "test description",
    "type": "MAMMALS",
  },
]
`;

exports[`AnimalsController getAllAnimals should throw an error when there is something wrong with mongodb connection 1`] = `[InternalServerErrorException: Internal Server Error]`;

exports[`AnimalsController getAnimal should return an animal 1`] = `
{
  "animalName": "Human",
  "createdAt": 2023-03-19T18:27:12.933Z,
  "description": "test description",
  "type": "MAMMALS",
}
`;

exports[`AnimalsController getAnimal should throw error when animal not found 1`] = `[NotFoundException: Not Found]`;
